keras_maskrcnn/preprocessing/generator.py

```diff
@@ -0,0 +1,278 @@
+
+    def group_images(self):
+        # determine the order of the images
+        order = list(range(self.size()))
+        if self.group_method == 'random':
+            random.shuffle(order)
+        elif self.group_method == 'ratio':
+            order.sort(key=lambda x: self.image_aspect_ratio(x))
+
+        # divide into groups, one group = one batch
+        self.groups = [[order[x % len(order)] for x in range(i, i + self.batch_size)] for i in range(0, len(order), self.batch_size)]
+
+    def compute_inputs(self, image_group):
+        # get the max image shape
+        max_shape = tuple(max(image.shape[x] for image in image_group) for x in range(3))
+
+        # construct an image batch object
+        image_batch = np.zeros((self.batch_size,) + max_shape, dtype=keras.backend.floatx())
+
+        # copy all images to the upper left part of the image batch object
+        for image_index, image in enumerate(image_group):
+            image_batch[image_index, :image.shape[0], :image.shape[1], :image.shape[2]] = image
+
+        return image_batch
+
+    def compute_targets(self, image_group, annotations_group, masks_group):
+        # get the max image shape
+        max_shape = tuple(max(image.shape[x] for image in image_group) for x in range(3))
+
+        # compute labels and regression targets
+        labels_group     = [None] * self.batch_size
+        regression_group = [None] * self.batch_size
+
+        for index, (image, annotations) in enumerate(zip(image_group, annotations_group)):
+            # compute regression targets
+            labels_group[index], annotations, anchors = self.compute_anchor_targets(
+                max_shape,
+                annotations,
+                self.num_classes(),
+                mask_shape=image.shape,
+            )
+            regression_group[index] = bbox_transform(anchors, annotations)
+
+            # append anchor states to regression targets (necessary for filtering 'ignore', 'positive' and 'negative' anchors)
+            anchor_states           = np.max(labels_group[index], axis=1, keepdims=True)
+            regression_group[index] = np.append(regression_group[index], anchor_states, axis=1)
+
+        labels_batch     = np.zeros((self.batch_size,) + labels_group[0].shape, dtype=keras.backend.floatx())
+        regression_batch = np.zeros((self.batch_size,) + regression_group[0].shape, dtype=keras.backend.floatx())
+
+        # copy all labels and regression values to the batch blob
+        for index, (labels, regression) in enumerate(zip(labels_group, regression_group)):
+            labels_batch[index, ...]     = labels
+            regression_batch[index, ...] = regression
+
+        # copy all annotations / masks to the batch
+        max_annotations = max(a.shape[0] for a in annotations_group)
+        masks_batch     = np.zeros((self.batch_size, max_annotations, 5 + 2 + max_shape[0] * max_shape[1]), dtype=keras.backend.floatx())
```

Comment:
Maybe document somewhere what this strange shape means.

Reply:
done.